app/controllers/agent/manage.go

```diff
@@ -98,6 +107,59 @@ func (c Agent) SendCmd(uuid string) revel.Result {
 	return c.RenderJSON(reply)
 }
 
+// addVisualExplain converts classic explain in JSON form into visual explain.
+func addVisualExplain(data []byte) ([]byte, error) {
+	explains := struct {
+		Classic []proto.ExplainRow
+		JSON    string
+		Visual  string
+	}{[]proto.ExplainRow{}, "", ""}
+	err := json.Unmarshal(data, &explains)
+	if err != nil {
+		return []byte{}, fmt.Errorf("cannot unmarshal classic expain to do visual explain: %s", err.Error())
+	}
+	rawExplainRows := []string{"id\tselect_type\ttable\tpartitions\ttype\tpossible_keys\tkey\tkey_len\tref\trows\tfiltered\tExtra"}
+	for _, explainRow := range explains.Classic {
+		id, _ := explainRow.Id.Value()
+		selectType, _ := explainRow.SelectType.Value()
+		table, _ := explainRow.Table.Value()
+		partitions, _ := explainRow.Partitions.Value()
+		theType, _ := explainRow.Type.Value()
+		possibleKeys, _ := explainRow.PossibleKeys.Value()
+		key, _ := explainRow.Key.Value()
+		keyLen, _ := explainRow.KeyLen.Value()
+		ref, _ := explainRow.Ref.Value()
+		rows, _ := explainRow.Rows.Value()
+		filtered, _ := explainRow.Filtered.Value()
+		extra, _ := explainRow.Extra.Value()
+		explainRowString := fmt.Sprintf("%v\t%v\t%v\t%v\t%v\t%v\t%v\t%v\t%v\t%v\t%v\t%v",
+			id,
+			selectType,
+			table,
+			partitions,
+			theType,
+			possibleKeys,
+			key,
+			keyLen,
+			ref,
+			rows,
+			filtered,
+			extra,
+		)
+		explainRowString = strings.Replace(explainRowString, "<nil>", "NULL", -1)
+		rawExplainRows = append(rawExplainRows, explainRowString)
+	}
+	rawExplain := strings.Join(rawExplainRows, "\n")
+
+	cmd := exec.Command("bash", "-c", "pt-visual-explain <(echo '"+rawExplain+"')")
```

Comment:
Are you sure this kind of escaping is enough for `rawExplain`? I was thinking, maybe passing as stdin would be safer and nicer? Can `pt-visual-explain` accept stdin (`mysql -e "explain select * from mysql.user" | pt-visual-explain`, there is pipe so looks like yes)? https://golang.org/pkg/os/exec/#Cmd.StdinPipe

Reply:
Yes, I have checked, this kind of escaping is enough for `rawExplain`. But, I'll double check, may be we even can remove these single quotes.